File: fadecandy-app/src/main/java/fr/bmartel/fadecandy/ledcontrol/Spark.java
package fr.bmartel.fadecandy.ledcontrol;


import android.graphics.Color;

import fr.bmartel.opc.Animation;
import fr.bmartel.opc.PixelStrip;

/**
 * Display a moving white pixel with trailing orange/red flames
 * This looks pretty good with a ring of pixels.
 */
public class Spark extends Animation {

    int color[] = null;

    int currentPixel;
    int numPixels;

    public Spark(int[] colors) {
        this.color = colors;
    }

    @Override
    public void reset(PixelStrip strip) {
        currentPixel = 0;
        numPixels = strip.getPixelCount();
    }

    @Override
    public boolean draw(PixelStrip strip) {
        for (int i = 0; i < color.length; i++) {
            strip.setPixelColor(pixNum(currentPixel, i), color[i]);
        }

        currentPixel = pixNum(currentPixel + 1, 0);
        return true;
    }

    public static int convertSpeed(int speed) {
        if (speed == 100) {
            return 5;
        } else {
            return (100 - speed);
        }
    }

    /**
     * Return the pixel number that is i steps behind number p.
     */
    int pixNum(int p, int i) {
        return (p + numPixels - i) % numPixels;
    }

    public static int[] buildColors(int color, int sparkSpan) {
        int red = Color.red(color);
        int green = Color.green(color);
        int blue = Color.blue(color);

        int fullColor = makeColor(red, green, blue);
        int noColor = makeColor(0, 0, 0);

        int[] colors = new int[sparkSpan + 1];

        for (int i = 0; i < colors.length - 1; i++) {
            colors[i] = fullColor;
        }
        colors[colors.length - 1] = noColor;

        return colors;
    }

    public void updateColor(int mColor, int sparkSpan) {
        color = buildColors(mColor, sparkSpan);
    }
}

File: fadecandy-app/src/main/java/fr/bmartel/fadecandy/ledcontrol/Spark.kt
package fr.bmartel.fadecandy.ledcontrol

import android.graphics.Color

import fr.bmartel.opc.Animation
import fr.bmartel.opc.PixelStrip

/**
 * Display a moving white pixel with trailing orange/red flames
 * This looks pretty good with a ring of pixels.
 */
class Spark(colors: IntArray) : Animation() {

    var color: IntArray
    private var currentPixel: Int = 0
    private var numPixels: Int = 0

    init {
        this.color = colors
    }

    override fun reset(strip: PixelStrip) {
        currentPixel = 0
        numPixels = strip.pixelCount
    }

    override fun draw(strip: PixelStrip): Boolean {
        for (i in color.indices) {
            strip.setPixelColor(pixNum(currentPixel, i), color[i])
        }

        currentPixel = pixNum(currentPixel + 1, 0)
        return true
    }

    /**
     * Return the pixel number that is i steps behind number p.
     */
    private fun pixNum(p: Int, i: Int): Int {
        return (p + numPixels - i) % numPixels
    }

    fun updateColor(mColor: Int, sparkSpan: Int) {
        color = buildColors(mColor, sparkSpan)
    }

    companion object {

        fun convertSpeed(speed: Int): Int {
            return if (speed == 100) {
                5
            } else {
                100 - speed
            }
        }

        fun buildColors(color: Int, sparkSpan: Int): IntArray {
            val red = Color.red(color)
            val green = Color.green(color)
            val blue = Color.blue(color)

            val fullColor = Animation.makeColor(red, green, blue)
            val noColor = Animation.makeColor(0, 0, 0)

            val colors = IntArray(sparkSpan + 1)

            for (i in 0 until colors.size - 1) {
                colors[i] = fullColor
            }
            colors[colors.size - 1] = noColor

            return colors
        }
    }
}